File: core/classify.py
"""
Layer 1 -- classification / structured extraction.

In production this is the "Hardware Domain Parser (LLM + Pydantic)" from
the architecture this project is based on: an LLM reads the raw listing and
fills the EngineeringJobSpec-shaped schema. Swapping in a real LLM here is
a one-function change -- see core.llm.LLMClient -- but the default path is
a deterministic keyword/heuristic extractor so the whole pipeline runs with
zero external dependencies and zero API cost.
"""
from __future__ import annotations

import re

from core.models import ChannelType, ClientArchetype, ManufacturingDomain, RawSignal

_CAD_TOOLS = ["Fusion 360", "SolidWorks", "Altium", "KiCad", "AutoCAD", "STEP", "IGES", "Creo", "CATIA"]
_MATERIALS = ["aluminum", "aluminium", "ABS", "PLA", "steel", "stainless steel", "nylon", "PC", "brass", "titanium"]

_DOMAIN_RULES: list[tuple[ManufacturingDomain, re.Pattern]] = [
    (ManufacturingDomain.PCB_ELECTRONICS, re.compile(r"\bpcb\b|\baltium\b|\bkicad\b|\bgerber\b|\bschematic\b|\bfirmware\b|\bembedded\b", re.I)),
    (ManufacturingDomain.DFM_INJECTION_MOLDING, re.compile(r"injection mold|\bdfm\b|draft angle|mold flow|tooling", re.I)),
    (ManufacturingDomain.SHEET_METAL, re.compile(r"sheet metal|stamping|bending|extrusion", re.I)),
    (ManufacturingDomain.PROTOTYPING_3D_PRINT, re.compile(r"3d print|\bsla\b|\bfdm\b|\bsls\b|rapid prototyp", re.I)),
    (ManufacturingDomain.ENCLOSURE_DESIGN, re.compile(r"enclosure|housing|casing", re.I)),
    (ManufacturingDomain.FULL_NPD_TURNKEY, re.compile(r"turnkey|napkin sketch|concept to production|full product development", re.I)),
    (ManufacturingDomain.CAD_MECHANICAL, re.compile(r"\bcad\b|solidworks|fusion ?360|mechanical design|\bstep\b file|\biges\b", re.I)),
]

_STAGE_RULES = [
    (re.compile(r"napkin sketch|just an idea|concept only|early concept", re.I), "Concept / Napkin Sketch"),
    (re.compile(r"working prototype|proof of concept|\bpoc\b", re.I), "Prototype / Proof of Concept"),
    (re.compile(r"ready for production|production ready|need dfm|scale to production", re.I), "Optimization / DFM"),
    (re.compile(r"mass production|high volume|going into production", re.I), "Production Ready"),
]

_ARCHETYPE_BY_CHANNEL = {
    ChannelType.FREELANCE: ClientArchetype.NPD_INNOVATOR,
    ChannelType.B2B_TRADE: ClientArchetype.MIDDLEMAN_RESELLER,
    ChannelType.COMMUNITY: ClientArchetype.NPD_INNOVATOR,
    ChannelType.BROKERAGE: ClientArchetype.SME_ENGINEERING,
    ChannelType.OUTBOUND: ClientArchetype.CROWDFUNDER,
}

_ARCHETYPE_OVERRIDES = [
    (re.compile(r"government|tender|rfp\b|defense|public sector|procurement", re.I), ClientArchetype.INSTITUTIONAL),
    (re.compile(r"reseller|white.?label|oem|odm|import(?:er)?|amazon (?:fba|seller)|shopify brand", re.I), ClientArchetype.MIDDLEMAN_RESELLER),
    (re.compile(r"kickstarter|indiegogo|crowdfund|our backers", re.I), ClientArchetype.CROWDFUNDER),
    (re.compile(r"engineering team|internal team|overflow|bandwidth|fractional engineer", re.I), ClientArchetype.SME_ENGINEERING),
]


def classify_domain(text: str) -> ManufacturingDomain:
    for domain, pattern in _DOMAIN_RULES:
        if pattern.search(text):
            return domain
    return ManufacturingDomain.UNKNOWN


def classify_archetype(signal: RawSignal) -> ClientArchetype:
    text = f"{signal.title}\n{signal.raw_text}"
    for pattern, archetype in _ARCHETYPE_OVERRIDES:
        if pattern.search(text):
            return archetype
    return _ARCHETYPE_BY_CHANNEL.get(signal.channel_type, ClientArchetype.UNKNOWN)


def extract_cad_software(text: str) -> list[str]:
    return sorted({tool for tool in _CAD_TOOLS if re.search(re.escape(tool), text, re.I)})


def extract_materials(text: str) -> list[str]:
    return sorted({m for m in _MATERIALS if re.search(re.escape(m), text, re.I)})


def extract_deliverables(text: str) -> list[str]:
    deliverables = []
    checks = {
        r"\bstep\b file": "STEP file",
        r"\bgerber\b": "Gerber/BOM",
        r"2d (?:production )?drawing": "2D Production Drawings",
        r"\bbom\b": "BOM",
        r"render": "Renders",
        r"moldflow": "Moldflow Analysis",
        r"prototype": "Physical Prototype",
    }
    for pat, label in checks.items():
        if re.search(pat, text, re.I):
            deliverables.append(label)
    return deliverables or ["STEP file", "2D Production Drawings"]


def classify_project_stage(text: str) -> str:
    for pattern, stage in _STAGE_RULES:
        if pattern.search(text):
            return stage
    return "Unspecified"

File: iechm-leadgen-engine-c/src/classify.c
#include "classify.h"
#include "strutil.h"
#include <stdio.h>

/* Sorted (Python `sorted()` over a set, ASCII order) so extraction order
 * matches the original without needing a separate sort/dedup step here --
 * each list has no duplicate entries, so appending in this fixed order
 * already yields the deduped, sorted result Python produces. */
static const char *CAD_TOOLS_SORTED[] = {
    "Altium", "AutoCAD", "CATIA", "Creo", "Fusion 360", "IGES", "KiCad", "STEP", "SolidWorks",
};
#define CAD_TOOLS_N 9

static const char *MATERIALS_SORTED[] = {
    "ABS", "PC", "PLA", "aluminium", "aluminum", "brass", "nylon", "stainless steel", "steel", "titanium",
};
#define MATERIALS_N 10

void extract_cad_software(const char *text, char out[LIST_MAX][LIST_ITEM_LEN], int *out_count) {
    int n = 0;
    for (int i = 0; i < CAD_TOOLS_N && n < LIST_MAX; i++) {
        if (has_plain(text, CAD_TOOLS_SORTED[i])) xcpy(out[n++], LIST_ITEM_LEN, CAD_TOOLS_SORTED[i]);
    }
    *out_count = n;
}

void extract_materials(const char *text, char out[LIST_MAX][LIST_ITEM_LEN], int *out_count) {
    int n = 0;
    for (int i = 0; i < MATERIALS_N && n < LIST_MAX; i++) {
        if (has_plain(text, MATERIALS_SORTED[i])) xcpy(out[n++], LIST_ITEM_LEN, MATERIALS_SORTED[i]);
    }
    *out_count = n;
}

void extract_deliverables(const char *text, char out[LIST_MAX][LIST_ITEM_LEN], int *out_count) {
    int n = 0;
    if (has_prefix_word(text, "step file") && n < LIST_MAX) xcpy(out[n++], LIST_ITEM_LEN, "STEP file");
    if (has_word(text, "gerber") && n < LIST_MAX) xcpy(out[n++], LIST_ITEM_LEN, "Gerber/BOM");
    if ((has_plain(text, "2d drawing") || has_plain(text, "2d production drawing")) && n < LIST_MAX)
        xcpy(out[n++], LIST_ITEM_LEN, "2D Production Drawings");
    if (has_word(text, "bom") && n < LIST_MAX) xcpy(out[n++], LIST_ITEM_LEN, "BOM");
    if (has_plain(text, "render") && n < LIST_MAX) xcpy(out[n++], LIST_ITEM_LEN, "Renders");
    if (has_plain(text, "moldflow") && n < LIST_MAX) xcpy(out[n++], LIST_ITEM_LEN, "Moldflow Analysis");
    if (has_plain(text, "prototype") && n < LIST_MAX) xcpy(out[n++], LIST_ITEM_LEN, "Physical Prototype");

    if (n == 0) {
        xcpy(out[0], LIST_ITEM_LEN, "STEP file");
        xcpy(out[1], LIST_ITEM_LEN, "2D Production Drawings");
        n = 2;
    }
    *out_count = n;
}

const char *classify_project_stage(const char *text) {
    if (has_plain(text, "napkin sketch") || has_plain(text, "just an idea") ||
        has_plain(text, "concept only") || has_plain(text, "early concept"))
        return "Concept / Napkin Sketch";

    if (has_plain(text, "working prototype") || has_plain(text, "proof of concept") || has_word(text, "poc"))
        return "Prototype / Proof of Concept";

    if (has_plain(text, "ready for production") || has_plain(text, "production ready") ||
        has_plain(text, "need dfm") || has_plain(text, "scale to production"))
        return "Optimization / DFM";

    if (has_plain(text, "mass production") || has_plain(text, "high volume") ||
        has_plain(text, "going into production"))
        return "Production Ready";

    return "Unspecified";
}

ManufacturingDomain classify_domain(const char *text) {
    if (has_word(text, "pcb") || has_word(text, "altium") || has_word(text, "kicad") ||
        has_word(text, "gerber") || has_word(text, "schematic") || has_word(text, "firmware") ||
        has_word(text, "embedded"))
        return DOM_PCB_ELECTRONICS;

    if (has_plain(text, "injection mold") || has_word(text, "dfm") || has_plain(text, "draft angle") ||
        has_plain(text, "mold flow") || has_plain(text, "tooling"))
        return DOM_DFM_INJECTION_MOLDING;

    if (has_plain(text, "sheet metal") || has_plain(text, "stamping") || has_plain(text, "bending") ||
        has_plain(text, "extrusion"))
        return DOM_SHEET_METAL;

    if (has_plain(text, "3d print") || has_word(text, "sla") || has_word(text, "fdm") ||
        has_word(text, "sls") || has_plain(text, "rapid prototyp"))
        return DOM_PROTOTYPING_3D_PRINT;

    if (has_plain(text, "enclosure") || has_plain(text, "housing") || has_plain(text, "casing"))
        return DOM_ENCLOSURE_DESIGN;

    if (has_plain(text, "turnkey") || has_plain(text, "napkin sketch") ||
        has_plain(text, "concept to production") || has_plain(text, "full product development"))
        return DOM_FULL_NPD_TURNKEY;

    if (has_word(text, "cad") || has_plain(text, "solidworks") ||
        has_plain(text, "fusion360") || has_plain(text, "fusion 360") ||
        has_plain(text, "mechanical design") || has_prefix_word(text, "step file") || has_word(text, "iges"))
        return DOM_CAD_MECHANICAL;

    return DOM_UNKNOWN;
}

static bool has_white_label(const char *text) {
    int pos = ci_find(text, "white", 0);
    while (pos >= 0) {
        int after = pos + 5;
        if (ci_find(text, "label", after) == after) return true;
        if (text[after] != '\0' && ci_find(text, "label", after + 1) == after + 1) return true;
        pos = ci_find(text, "white", pos + 1);
    }
    return false;
}

static ClientArchetype archetype_override(const char *text) {
    if (has_plain(text, "government") || has_plain(text, "tender") || has_suffix_word(text, "rfp") ||
        has_plain(text, "defense") || has_plain(text, "public sector") || has_plain(text, "procurement"))
        return ARCH_INSTITUTIONAL;

    if (has_plain(text, "reseller") || has_white_label(text) || has_plain(text, "oem") ||
        has_plain(text, "odm") || has_plain(text, "import") ||
        has_plain(text, "amazon fba") || has_plain(text, "amazon seller") || has_plain(text, "shopify brand"))
        return ARCH_MIDDLEMAN_RESELLER;

    if (has_plain(text, "kickstarter") || has_plain(text, "indiegogo") ||
        has_plain(text, "crowdfund") || has_plain(text, "our backers"))
        return ARCH_CROWDFUNDER;

    if (has_plain(text, "engineering team") || has_plain(text, "internal team") ||
        has_plain(text, "overflow") || has_plain(text, "bandwidth") || has_plain(text, "fractional engineer"))
        return ARCH_SME_ENGINEERING;

    return ARCH_UNKNOWN; /* sentinel meaning "no override matched" */
}

static ClientArchetype archetype_by_channel(ChannelType ct) {
    switch (ct) {
        case CT_FREELANCE: return ARCH_NPD_INNOVATOR;
        case CT_B2B_TRADE: return ARCH_MIDDLEMAN_RESELLER;
        case CT_COMMUNITY: return ARCH_NPD_INNOVATOR;
        case CT_BROKERAGE: return ARCH_SME_ENGINEERING;
        case CT_OUTBOUND: return ARCH_CROWDFUNDER;
        default: return ARCH_UNKNOWN;
    }
}

ClientArchetype classify_archetype(const RawSignal *signal) {
    char text[TITLE_LEN + TEXT_LEN + 2];
    snprintf(text, sizeof text, "%s\n%s", signal->title, signal->raw_text);

    ClientArchetype override = archetype_override(text);
    if (override != ARCH_UNKNOWN) return override;
    return archetype_by_channel(signal->channel_type);
}
